src/org/openstreetmap/josm/gui/NavigatableComponent.java

```diff
@@ -149,6 +154,51 @@ protected static void fireZoomChanged() {
         });
     }
 
+    private final CopyOnWriteArrayList<PrimitiveHoverListener> primitiveHoverListeners = new CopyOnWriteArrayList<>();
+    // Allow primitive to be garbage collected by using a weak reference
+    private WeakReference<OsmPrimitive> previousHoveredPrimitive = new WeakReference<>(null);
+    private PrimitiveHoverMouseListener primitiveHoverMouseListenerHelper = new PrimitiveHoverMouseListener();
+
+    /**
+     * Removes a primitive hover listener
+     *
+     * @param listener the listener. Ignored if null or already absent.
+     */
+    public void removePrimitiveHoverListener(PrimitiveHoverListener listener) {
+        primitiveHoverListeners.remove(listener);
+    }
+
+    /**
+     * Adds a primitive hover listener
+     *
+     * @param listener the listener. Ignored if null or already registered.
+     */
+    public void addPrimitiveHoverListener(PrimitiveHoverListener listener) {
+        if (listener != null) {
+            primitiveHoverListeners.addIfAbsent(listener);
+        }
+    }
+
+    protected void firePrimitiveHovered(PrimitiveHoverEvent e) {
+        GuiHelper.runInEDT(() -> {
+            for (PrimitiveHoverListener l : primitiveHoverListeners) {
+                try {
+                    l.primitiveHovered(e);
+                } catch (Exception ex) {
+                    Logging.error(ex);
+                }
```

Comment:
Exceptions are caught here to avoid a situation like this https://josm.openstreetmap.de/ticket/21682
I expect that quite a few plugins might want to use this listener and an error in a plugin should not interfere with the operation of the rest of the application.

Reply:
Both true and false. I'd rather force plugins to fix their code than have plugins _think_ something is working when it isn't. How often do people actually read all the output from the `Logging` methods?

If it turns out to block JOSM shutdown, we can re-evaluate.

Reply:
I changed it to pop the bug report window (which is what I wanted to do in the first place but somehow thought the error logging already did)